assignments/solutions/fill_in_the_blanks/src/closures.rs

```diff
@@ -0,0 +1,71 @@
+struct BaleOfStraw {
+    size: usize,
+}
+
+fn main() {
+    // Recap: Anatomy of a closure
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    //
+    let rummage = |element| {
+        // 👈  input parameters MAY omit type annotation
+        //     if type can be inferred. The input parameter
+        //     is also referred to as the "captured value".
+        println!("found: {}", element);
+        // 👈  no return statement: closures *can* return values,
+        //     but this one doesn't
+    }; // 👈  {}s are only needed for multi-line closures
+
+    let haystack = vec!["hay", "hay", "hay", "needle", "hay", "hay"];
+
+    // REFERENCING CLOSURES
+    //=====================
+    // *By default*, closures capture variables by reference. When given a choice, this is the
+    // behaviour inferred by the compiler.
+
+    haystack.iter().for_each(rummage);
+    // note that we can directly borrow the contents of `haystack` after using the `rummage`
+    // closure, because it only borrowed the elements (and is now done)
+    println!("top of the haystack: {}", haystack[0]);
+
+    // MUTATING CLOSURES
+    //==================
+    // Closures can mutate the variables they are capturing
+
+    // 👀  Closures can be used as function arguments.
+
+    // ✅ TODO: remove all the hay from `haystack` by checking whether `key` is a needle
+    let mut haystack_clone = haystack.clone();
+    haystack_clone.retain(|key| *key == "needle");
+    println!("look, I found the needle: {:?}", haystack_clone);
+
+    // 👀  a common use case for closures is to transform collections
+    //     using e.g. `map()` and `filter()`.
+
+    // ✅ TODO: use `map()` to increase the bale size with every "hay" in the haystack
+    let mut bale = BaleOfStraw { size: 0 };
+
+    let empty_haystack: Vec<_> = haystack
+        .into_iter()
+        .filter(|element| *element == "hay")
+        .map(|_| bale.size += 1)
```

Comment:
how about this use of `map`?

```rust
    let haystack = vec!["hay", "hay", "hay", "needle", "hay", "hay"];
    // ...
    let heystack: Vec<_> = haystack
        .into_iter()
        .filter(|element| *element == "hay")
        .map(|item| item.replace("hay", "hey"))
        .collect();
    println!("hey! {:?}", heystack);
```

Reply:
haha, I like the play on words a lot! Unfortunately I think the difference between "hay" and "hey" is easy to miss during the frenzy of a training so I will sliiightly modify it though